staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -0,0 +1,458 @@
+		k8serrors.IsUnsupportedMediaType(e) ||
+		k8serrors.IsMethodNotSupported(e)
+}
+
+func (v *Validator) recordConflicts(candidate *subnet.Summary, potentialRivals []netv1a1.Subnet) ([]string, bool, error) {
+	var conflictsMsgs []string
+	var conflictFound bool
+
+	for _, pr := range potentialRivals {
+		potentialRival, parsingErrs := subnet.NewSummary(&pr)
+		if len(parsingErrs) > 0 {
+			glog.Errorf("parsing %s failed while validating %s: %s", potentialRival.NamespacedName, candidate.NamespacedName, parsingErrs.Error())
+		}
+
+		if !potentialRival.Conflict(candidate) || potentialRival.SameSubnetAs(candidate) {
+			// potentialRival is not a rival to candidate or it is the same
+			// subnet as candidate, hence we skip it.
+			continue
+		}
+
+		// If we're here the two subnets represented by potentialRival and
+		// candidate are in conflict, that is, they are rivals.
+		conflictFound = true
+		if potentialRival.CIDRConflict(candidate) {
+			glog.V(2).Infof("CIDR conflict found between %s (%d, %d) and %s (%d, %d).", candidate.NamespacedName, candidate.BaseU, candidate.LastU, potentialRival.NamespacedName, potentialRival.BaseU, potentialRival.LastU)
+			conflictsMsgs = append(conflictsMsgs, fmt.Sprintf("CIDR overlaps with %s's (%s)", potentialRival.NamespacedName, pr.Spec.IPv4))
+		}
+		if potentialRival.NSConflict(candidate) {
+			glog.V(2).Infof("Namespace conflict found between %s and %s.", candidate.NamespacedName, potentialRival.NamespacedName)
+			conflictsMsgs = append(conflictsMsgs, fmt.Sprintf("same VNI but different namespace wrt %s", potentialRival.NamespacedName))
+		}
+
+		// Record the conflict in the conflicts cache.
+		if err := v.recordConflict(potentialRival, candidate); err != nil {
+			return conflictsMsgs, conflictFound, err
+		}
+	}
+
+	return conflictsMsgs, conflictFound, nil
+}
+
+func (v *Validator) updateSubnetValidity(s *netv1a1.Subnet, validated bool, validationErrors []string) error {
+	sCopy := s.DeepCopy()
+
+	sCopy.Status.Validated = validated
+	sCopy.Status.Errors.Validation = validationErrors
+
+	_, err := v.netIfc.Subnets(sCopy.Namespace).Update(sCopy)
+	switch {
+	case err == nil:
+		nsn := k8stypes.NamespacedName{
+			Namespace: s.Namespace,
+			Name:      s.Name,
+		}
+		v.updateStaleRV(nsn, s.ResourceVersion)
+	case doNotRetryUpdate(err):
+		glog.Errorf("failed to update subnet from %#+v to %#+v: %s. There will be no retry because of the nature of the error", s, sCopy, err.Error())
+	case !doNotRetryUpdate(err):
+		return fmt.Errorf("failed to update subnet from %#+v to %#+v: %s", s, sCopy, err.Error())
+	}
+
+	return nil
+}
+
+func (v *Validator) recordConflict(enroller, enrollee *subnet.Summary) error {
+	v.conflictsMutex.Lock()
+	defer v.conflictsMutex.Unlock()
+
+	c := v.conflicts[enroller.NamespacedName]
+
+	if c == nil {
+		return fmt.Errorf("registration of %s as a rival of %s failed: %s's conflicts cache not found", enrollee.NamespacedName, enroller.NamespacedName, enroller.NamespacedName)
```

Comment:
Huh?  I think we need to add a conflict cache in this case.  Consider that at startup, the first time any subnet is processed we may find a conflict and need to record it.

Reply:
No, that code is right. A conflicts cache should be initialized only by the owner of the cache, not by a subnet in conflict with the owner (as you're proposing). To avoid race conditions due to concurrent validation of a subnet X and updates to VNI and CIDR to a subnet Y in conflict with X, you need to compare Y's VNI and CIDR to those recorded in its conflicts cache. The code you commented serves the same purpose, but it avoids the race condition due to concurrent validation of a subnet X and deletion of a subnet Y in conflict with X. Otherwise X is added to Y's conflicts cache even if Y has been deleted! If Y was the only conflict with X this would prevent X from being validated.

> Consider that at startup, the first time any subnet is processed we may find a conflict and need to record it

That is not a problem. During processing of an existing subnet X its conflicts cache is initialized by [updateConflictsCache](https://github.com/MikeSpreitzer/kube-examples/blob/make-subnets-validation-correct/staging/kos/pkg/controllers/subnet/validator.go#L245) before [looking for conflicts](https://github.com/MikeSpreitzer/kube-examples/blob/make-subnets-validation-correct/staging/kos/pkg/controllers/subnet/validator.go#L281).

Reply:
Oh, you're right, there is a race with deletion.  Another way to see the correctness of the existing code is to observe that it is the extension of the `if !enroller.Equal(c.ownerSummary) {...}` statement to cover the case where c is not reified yet.

Also, since this returns an error that will cause a retry, this is safe.

The fact that updateConflictsCache(X) is called before recordConflicts(X, ...) is not what solves the problem of recordConflicts(X, ...) finding a conflict with Y before any worker ever called updateConflictsCache(Y).